### PR TITLE
Partial Merkle tree implementation

### PR DESCRIPTION
**Pull request by Overcastan**
Implementation of basic functionality for Partial Merkle tree similar to Sparse Merkle tree.
(De)serialization of this structure will be implemented in future PR.

### CODE REVIEW

File: src/merkle/partial_mt/mod.rs [hunk 1]
```diff
@@ -0,0 +1,270 @@
...
+        }
+
+        for (node_index, rpo_digest) in entries {
+            let old_value = tree.update_leaf(node_index, rpo_digest)?;
+            if old_value != Self::EMPTY_DIGEST {
+                return Err(MerkleError::DuplicateValuesForIndex(node_index.value()));
+            }
+        }
+        Ok(tree)
+    }
```
Comment (bobbinth) [inline src/merkle/partial_mt/mod.rs:85]: We should not be able to create a `PartialMerkleTree` from leaves. If we are able to do so, then it becomes a regular Merkle tree. What we want is to create a partial Merkle tree from Merkle paths (similar to [this](https://github.com/0xPolygonMiden/crypto/blob/main/src/merkle/path_set.rs#L33) method in `MerklePathSet`). I would change this to something like:
```Rust
pub fn with_paths<I>(paths: I) -> Result<Self, MerkleError>
where
    I: IntoIterator<Item = (u64, Word, MerklePath)>,
{
    ...
}
```
The main difference between this is `MerklePathSet` is that for `PartialMerkleTree` paths could be of different lengths.

In the future, we can add other constructors - but I think for now this is all we need.

File: src/merkle/partial_mt/mod.rs [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+    /// Returns the root of this Merkle tree.
+    pub fn root(&self) -> Word {
+        self.root.into()
+    }
+
+    /// Returns the depth of this Merkle tree.
+    // TODO: maybe it's better to rename it to the `max_depth`
+    pub fn depth(&self) -> u8 {
+        self.max_depth
+    }
```
Comment (bobbinth) [inline src/merkle/partial_mt/mod.rs:99]: Yes - I'd change this to `max_depth`.

File: src/merkle/partial_mt/mod.rs [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+
+    /// Returns a node at the specified NodeIndex.
+    ///
+    /// # Errors
+    /// Returns an error if the specified NodeIndex is not contained in the nodes map.
+    pub fn get_node(&self, index: NodeIndex) -> Result<Word, MerkleError> {
```
Comment (bobbinth) [inline src/merkle/partial_mt/mod.rs:105]: let's return a `RpoDigest` from here instead of `Word`.

File: src/merkle/partial_mt/mod.rs [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+                depth: index.depth(),
+                value: index.value(),
+            })
+        } else {
+            self.nodes
+                .get(&index)
+                .ok_or(MerkleError::NodeNotInSet(index))
+                .map(|hash| **hash)
+        }
+    }
```
Comment (bobbinth) [inline src/merkle/partial_mt/mod.rs:129]: I'm actually not sure if we need this method. But we could add something like:
```Rust
pub fn is_leaf(&self, index: NodeIndex) -> bool
```

Also, probably something like [get_leaf_depth()](https://github.com/0xPolygonMiden/crypto/blob/main/src/merkle/store/mod.rs#L203) would be useful. The signature could be:
```Rust
pub fn get_leaf_depth(&self, index: u64) -> Result<u8, MerkleError> {
    ...
}
```

**Review by bobbinth — CHANGES_REQUESTED** [review]
Thank you! This is not a full review yet - but I left some comments which will probably lead to some significant changes - so, I can review other parts once these are implemented.

File: src/merkle/partial_mt/mod.rs (hunk 1 above)
Comment (Overcastan) [reply src/merkle/partial_mt/mod.rs:85]: Got it, I'll create `with_paths` constructor. But maybe we should keep `with_leaves` one, because it is possible to pass "leaves" with different depths into it (so they're not actually leaves, more like outer nodes). So, for example, these two trees are equal:
```
with_paths([((2, 0), A, [(2, 1), (1, 1)]),
            ((3, 2), B, [(3, 3), (2, 0), (1, 1)]),
            ((3, 3), C, [(3, 2), (2, 0), (1, 1)]),
            ((1, 1), D, [(1, 0)]),
])
with_leaves([((2, 0), A), ((3, 2), B), ((3, 3), C), ((1, 1), D)])
```
since they both construct this tree:
```
                _______ (0, 0) _______
               /                      \
         _ (1, 0) _               _____ D _____
        /          \             /             \
    _ A _        (2, 1)       (2, 2)         (2, 3)
   /     \        /  \       /     \         /     \
(3, 0) (3, 1)    B    C   (3, 4) (3, 5)   (3, 6) (3, 7)
```
But on the other hand, Merkle paths are more common way to create trees, so, maybe, we just don't need `with_leaves` constructor.

Comment (bobbinth) [reply src/merkle/partial_mt/mod.rs:85]: I think with `with_leaves()` could work too - but we need to make sure we detect missing entries. For example, for the above tree, `with_leaves([((2, 0), A), ((3, 2), B), ((3, 3), C)])` should be an error (because we don't know what is under the right subtree of the root.

We also need to make sure that no duplicate entries are provided. For example, `with_leaves([((2, 0), A), ((3, 2), B), ((3, 3), C), ((1, 1), D), ((2, 2), F)])` should be an error (because `D` is no longer a leaf. Though, I think maybe this check is already happening.

Basically, I wonder if `with_leaves()` is a bit more complicated and something we can do in a follow-up PR.

File: src/merkle/partial_mt/mod.rs (hunk 4 above)
Comment (Overcastan) [reply src/merkle/partial_mt/mod.rs:129]: I'm not sure I understand how `get_leaf_depth` works -- what index is provided into it? Is it a "global" index of the node in the tree? Not sure how this function can be used in case of `PartialMerkleTree`.

Comment (bobbinth) [reply src/merkle/partial_mt/mod.rs:129]: `get_leaf_depth` would traverse the tree from the root down according to the provided index and would stop at the first leaf it finds.

Let's say this is our Partial Merkle tree:
```
                _______ (0, 0) _______
               /                      \
         _ (1, 0) _                    D
        /          \             
       A         (2, 1)    
                  /  \  
                 B    C
```
Calling `get_leaf_depth(0)` would return the depth of `A` (i.e., 2). But also calling `get_leaf_dpeth(1)` would return the depth of `A`.

Calling `get_leaf_depth(2)` would return the depth of `B` (i.e., 3)`, and calling `get_leaf_depth(5)` would return the depth of `D` (i.e., 1).

File: src/merkle/partial_mt/mod.rs [hunk 5]
```diff
@@ -0,0 +1,365 @@
...
+        if !self.nodes.contains_key(&index) {
+            return Err(MerkleError::NodeNotInSet(index));
+        }
+
+        let mut path = Vec::new();
+        for _ in 0..index.depth() {
+            let sibling_index = Self::get_sibling_index(&index)?;
+            index.move_up();
+            let sibling_hash =
+                self.nodes.get(&sibling_index).cloned().unwrap_or(Self::EMPTY_DIGEST);
```
Comment (bobbinth) [inline src/merkle/partial_mt/mod.rs:163]: Why are we using `EMPTY_DIGEST` as default here? If a node is in the tree, then its path should aways be in the tree as well, no? So, the situation when sibling is not in the tree should never arise. If so, we should just use `.expect()` here.

Also, minor nit: I'd name the variable just `sibling` (rather than `sibling_hash`).

File: src/merkle/partial_mt/mod.rs [hunk 6]
```diff
@@ -0,0 +1,365 @@
...
+
+    // ITERATORS
+    // --------------------------------------------------------------------------------------------
+
+    /// Returns an iterator over the leaves of this [PartialMerkleTree].
+    pub fn leaves(&self) -> impl Iterator<Item = (NodeIndex, &Word)> {
+        self.nodes
+            .iter()
+            .filter(|(index, _)| self.leaves.contains(index))
+            .map(|(index, hash)| (*index, &(**hash)))
```
Comment (bobbinth) [inline src/merkle/partial_mt/mod.rs:177]: Wouldn't it be easier to iterate over `leaves` and for each leaf index look up the relevant node?

File: src/merkle/partial_mt/mod.rs [hunk 7]
```diff
@@ -0,0 +1,365 @@
...
+                .expect("Failure to get right child index");
+            let left_hash = self.nodes.get(&left_index).cloned().unwrap_or(Self::EMPTY_DIGEST);
+            let right_hash = self.nodes.get(&right_index).cloned().unwrap_or(Self::EMPTY_DIGEST);
+            InnerNodeInfo {
+                value: **digest,
+                left: *left_hash,
+                right: *right_hash,
+            }
+        })
+    }
```
Comment (bobbinth) [inline src/merkle/partial_mt/mod.rs:196]: Let's take this out of this PR and move it to a subsequent PR (unless we really need it here).

File: src/merkle/partial_mt/mod.rs [hunk 8]
```diff
@@ -0,0 +1,365 @@
...
+    /// Returns an error if:
+    /// - The depth of the specified node_index is greater than 64 or smaller than 1.
+    /// - The specified path is not consistent with other paths in the set (i.e., resolves to a
+    ///   different root).
+    pub fn add_path(
+        &mut self,
+        index_value: NodeIndex,
+        value: Word,
+        mut path: MerklePath,
+    ) -> Result<(), MerkleError> {
```
Comment (bobbinth) [inline src/merkle/partial_mt/mod.rs:214]: Should the type of `index_value` be `u64`? The the length of the path gives the information about the depth.

File: src/merkle/partial_mt/mod.rs [hunk 9]
```diff
@@ -0,0 +1,365 @@
...
+        let sibling_node_index = Self::get_sibling_index(&index_value)?;
+        self.leaves.insert(sibling_node_index);
+
+        // add first two nodes to the nodes map
+        self.nodes.insert(index_value, value.into());
+        self.nodes.insert(sibling_node_index, path[0].into());
+
+        // update the current path
+        let parity = index_value.value() & 1;
+        path.insert(parity as usize, value);
```
Comment (bobbinth) [inline src/merkle/partial_mt/mod.rs:229]: I would avoid modifying the path. I think we need this only to compute the initial `root` (I'd probably rename this to `node`). If so, I'd just use `NodeIndex::buildNode` to give the correct arrangement of the node and its sibling.

File: src/merkle/partial_mt/mod.rs [hunk 10]
```diff
@@ -0,0 +1,365 @@
...
+            self.nodes.insert(sibling_node, hash.into());
+
+            Rpo256::merge(&index_value.build_node(root, hash.into()))
+        });
+
+        let old_root = self.nodes.get(&Self::ROOT_INDEX).cloned().unwrap_or(Self::EMPTY_DIGEST);
+
+        // if the path set is empty (the root is all ZEROs), set the root to the root of the added
+        // path; otherwise, the root of the added path must be identical to the current root
+        if old_root == Self::EMPTY_DIGEST {
```
Comment (bobbinth) [inline src/merkle/partial_mt/mod.rs:253]: Doesn't `self.root()` already kind of do this? So, we should be able to say something like:
```Rust
if self.root() == EMPTY_DIGEST {
```

File: src/merkle/partial_mt/mod.rs [hunk 11]
```diff
@@ -0,0 +1,365 @@
...
+            self.nodes.insert(Self::ROOT_INDEX, root);
+        } else if old_root != root {
+            return Err(MerkleError::ConflictingRoots([*old_root, *root].to_vec()));
+        }
+
+        self.update_leaves()?;
```
Comment (bobbinth) [inline src/merkle/partial_mt/mod.rs:259]: Seems like this may be a very expensive call. Is there a way to figure out if any of the leaves became internal nodes while we are in the loop above?

File: src/merkle/partial_mt/mod.rs [hunk 12]
```diff
@@ -0,0 +1,365 @@
...
+            self.nodes.insert(index_value, root);
+
+            let sibling_node = Self::get_sibling_index_unchecked(&index_value);
+            // assume for now that all path nodes are leaves and add them to the leaves set
+            self.leaves.insert(sibling_node);
+
+            // insert node from Merkle path to the nodes map
+            self.nodes.insert(sibling_node, hash.into());
+
+            Rpo256::merge(&index_value.build_node(root, hash.into()))
```
Comment (bobbinth) [inline src/merkle/partial_mt/mod.rs:246]: When we insert nodes into `self.nodes` we get back the previous value at the same index. Is there a way to use this information somehow to determine what is a leaf and what is not?

File: src/merkle/partial_mt/tests.rs [hunk 13]
```diff
@@ -0,0 +1,221 @@
...
+    let value22 = int_to_node(22).into();
+    let value23 = int_to_node(23).into();
+
+    let value21 = Rpo256::merge(&[value32, value33]);
+    let value10 = Rpo256::merge(&[value20, value21]);
+    let value11 = Rpo256::merge(&[value22, value23]);
+
+    let path_33 = vec![*value32, *value20, *value11];
+
+    let path_22 = vec![*value23, *value10];
```
Comment (bobbinth) [inline src/merkle/partial_mt/tests.rs:62]: Instead of doing this by hand, I wonder if it would be better to test against `MerkleStore` as it also has the ability to handle paths of different depths. (this is applicable to most tests in this file).

File: src/merkle/partial_mt/mod.rs [hunk 14]
```diff
@@ -0,0 +1,365 @@
...
+            leaves: BTreeSet::new(),
+        }
+    }
+
+    /// Appends the provided paths iterator into the set.
+    ///
+    /// Analogous to [Self::add_path].
+    pub fn with_paths<I>(paths: I) -> Result<Self, MerkleError>
+    where
+        I: IntoIterator<Item = (NodeIndex, Word, MerklePath)>,
```
Comment (bobbinth) [inline src/merkle/partial_mt/mod.rs:59]: Should we use just `u64` instead of `NodeIndex`? The `MerklePath` already provides the information about the depth.

**Review by bobbinth — CHANGES_REQUESTED** [review]
Thank you! Looks good! I left some more comments inline.

File: src/merkle/partial_mt/tests.rs (hunk 13 above)
Comment (Overcastan) [reply src/merkle/partial_mt/tests.rs:62]: I don't think that it will be easier. If I correctly understood how to create a `MerkleStore` instance, first we need to create paths, which then will be used to create a `MerkleStore`, so we need to calculate most of these nodes anyway. `MerkleStore` can help to check correctness of `PartialMerkleTree` functions output, but that seems to be useless since we already calculated all necessary nodes and values to create `MerkleStore` and `PartialMerkleTree`. Or maybe I just don't understand how we can use `MerkleStore` here.

File: src/merkle/partial_mt/mod.rs [hunk 15]
```diff
@@ -0,0 +1,318 @@
...
+            }
+
+            let sibling_node = index_value.sibling();
+            // node became a leaf only if it is a new node (it wasn't in nodes map)
+            if !self.nodes.contains_key(&sibling_node) {
+                self.leaves.insert(sibling_node);
+            }
+
+            // insert node from Merkle path to the nodes map
+            self.nodes.insert(sibling_node, hash.into());
```
Comment (bobbinth) [inline src/merkle/partial_mt/mod.rs:228]: Could this be simplified as?:
```Rust
if self.nodes.insert(sibling_node, hash.into()).is_none() {
    self.leaves.insert(sibling_node);
}
```
Also, might be good to leave a short comment explaining why this works.

File: src/merkle/partial_mt/mod.rs [hunk 16]
```diff
@@ -0,0 +1,318 @@
...
+        // if the path set is empty (the root is all ZEROs), set the root to the root of the added
+        // path; otherwise, the root of the added path must be identical to the current root
+        if self.root() == EMPTY_DIGEST {
+            self.nodes.insert(ROOT_INDEX, root);
+        } else if self.root() != root {
+            return Err(MerkleError::ConflictingRoots([*self.root(), *root].to_vec()));
+        }
+
+        // self.update_leaves()?;
+
```
Comment (bobbinth) [inline src/merkle/partial_mt/mod.rs:242]: minor nit: let's remove this.

File: src/merkle/partial_mt/mod.rs [hunk 17]
```diff
@@ -0,0 +1,318 @@
...
+                    value,
+                    self.nodes
+                        .get(&right_index)
+                        .cloned()
+                        .ok_or(MerkleError::NodeNotInSet(right_index))?,
+                )
+            };
+            node_index.move_up();
+            value = Rpo256::merge(&[left, right]);
+            self.nodes.insert(node_index, value);
```
Comment (bobbinth) [inline src/merkle/partial_mt/mod.rs:294]: We should be able to simplify this quite a bit. Maybe something like:
```Rust
let sibling = self.nodes.get(&node_index.sibling()).expect("sibling should exist");
value = Rpo256::merge(&node_index.build_node(value, *sibling));
node_index.move_up();
self.nodes.insert(node_index, value);
```

File: src/merkle/partial_mt/mod.rs [hunk 18]
```diff
@@ -0,0 +1,318 @@
...
+        // insert NodeIndex to the leaves Set
+        self.leaves.insert(node_index);
+
+        // add node value to the nodes Map
+        let old_value = self.nodes.insert(node_index, value).unwrap_or(EMPTY_DIGEST);
+
+        // if the old value and new value are the same, there is nothing to update
+        if value == old_value {
+            return Ok(value);
+        }
```
Comment (bobbinth) [inline src/merkle/partial_mt/mod.rs:267]: It is not clear to me why we are defaulting to `EMPTY_DIGEST` here. If we are trying to insert an `EMPTY_DIGEST` as a leaf, wouldn't this cause incorrect behavior?

File: src/merkle/partial_mt/tests.rs (hunk 13 above)
Comment (bobbinth) [reply src/merkle/partial_mt/tests.rs]: I guess it would depend on how we do it. We could create a `MerkleTree` (let's say of depth 3 or 4) and then create a `MerkleStore` from this Merkle tree. Then, we could get Merkle paths of various depths from this `MerkleStore` and create `PartialMerkleTree` from it. Then we could update the data in the Merkle store and make sure that similar updates to the Partial Merkle tree would give the same results.

Basically, I'm trying to avoid creating Merkle paths manually because we already have (tested) code which does that.

**Review by bobbinth — CHANGES_REQUESTED** [review]
Thank you! Looks good! I left some more comments inline.

File: src/merkle/partial_mt/mod.rs [hunk 19]
```diff
@@ -0,0 +1,318 @@
...
+    pub fn get_leaf_depth(&self, index: u64) -> Result<u8, MerkleError> {
+        let mut node_index = NodeIndex::new(self.max_depth(), index)?;
+        for _ in 0..node_index.depth() {
+            if self.leaves.contains(&node_index) {
+                return Ok(node_index.depth());
+            }
+            node_index.move_up()
+        }
+        Ok(0_u8)
+    }
```
[the remark this answers is not in the document]
Comment (Overcastan) [reply src/merkle/partial_mt/mod.rs:110]: It'll fail if passed index is greater than `2^max_depth`. For example, for a tree with `max_depth = 3` we can't pass an index greater than 7. Or did I misunderstand how this function should work?

File: src/merkle/partial_mt/mod.rs (hunk 17 above)
Comment (Overcastan) [reply src/merkle/partial_mt/mod.rs:294]: Ah, indeed! I always forget that we have a lot of convenient functions in the `NodeIndex` struct.

File: src/merkle/partial_mt/mod.rs (hunk 18 above)
Comment (Overcastan) [reply src/merkle/partial_mt/mod.rs:267]: Yes, my bad. I forgot to update this code from the moment when I mistakenly thought that all empty nodes were taken as zero. In that case, should it return `Result<Option<RpoDigest>, MerkleError>` instead of `Result<RpoDigest, MerkleError>`, since old value can be `None`?

File: src/merkle/partial_mt/mod.rs (hunk 19 above)
[the remark this answers is not in the document]
Comment (bobbinth) [reply src/merkle/partial_mt/mod.rs:111]: Ah - I see. I actually think maybe we don't need to enforce this. Meaning, we can truncate `index` so that it is appropriate for `max_depth`. This way, `get_leaf_depth` would be infallible.

File: src/merkle/partial_mt/mod.rs [hunk 20]
```diff
@@ -0,0 +1,318 @@
...
+                *leaf,
+                self.get_node(*leaf).unwrap_or_else(|_| {
+                    panic!(
+                        "Leaf with node index ({}, {}) is not in the nodes map",
+                        leaf.depth(),
+                        leaf.value()
+                    )
+                }),
+            )
+        })
```
[the remark this answers is not in the document]
Comment (bobbinth) [reply src/merkle/partial_mt/mod.rs]: Hmm - it doesn't seem to complain for me. Could you do `rustup update` and see if you still get this error?

File: src/merkle/partial_mt/mod.rs [hunk 21]
```diff
@@ -0,0 +1,306 @@
...
+
+    /// Returns true if provided index contains in the leaves set, false otherwise.
+    pub fn is_leaf(&self, index: NodeIndex) -> bool {
+        self.leaves.contains(&index)
+    }
+
+    pub fn get_leaf_depth(&self, index: u64) -> u8 {
+        let trunc_value = index.min(2_u64.pow(self.max_depth() as u32));
+        let mut node_index =
+            NodeIndex::new(self.max_depth(), trunc_value).expect("Truncated value is not valid.");
```
Comment (bobbinth) [inline src/merkle/partial_mt/mod.rs:108]: I'd do it differently here: first we can instantiate node index using `depth = 64`, and then we can use `NodeIndex::move_up_to()` to move to `max_depth`.

File: src/merkle/partial_mt/mod.rs [hunk 22]
```diff
@@ -0,0 +1,306 @@
+use super::{
+    BTreeMap, BTreeSet, MerkleError, MerklePath, NodeIndex, Rpo256, RpoDigest, ValuePath, Vec,
+    Word, EMPTY_WORD,
+};
+
+extern crate alloc;
+use alloc::format;
```
Comment (bobbinth) [inline src/merkle/partial_mt/mod.rs:7]: I'd use conditional imports similar to how it is done [here](https://github.com/facebook/winterfell/blob/main/utils/core/src/collections.rs#L12). Also, I'd move this to the utils module so that we could use it in other places as well if we need to.

File: src/merkle/partial_mt/tests.rs [hunk 23]
```diff
@@ -0,0 +1,128 @@
...
+        (3_u64, expected_path33.value.into(), expected_path33.path.clone()),
+        (2, expected_path22.value.into(), expected_path22.path.clone()),
+    ])
+    .unwrap();
+
+    let path33 = pmt.get_path(NODE33).unwrap();
+    let path22 = pmt.get_path(NODE22).unwrap();
+
+    assert_eq!(expected_path33.path, path33);
+    assert_eq!(expected_path22.path, path22);
```
Comment (bobbinth) [inline src/merkle/partial_mt/tests.rs:67]: I would add an `assert!` here to make sure the root of PMT is the same as the expected root.

File: src/merkle/partial_mt/tests.rs [hunk 24]
```diff
@@ -0,0 +1,128 @@
...
+    let path33 = ms.get_path(expected_root, NODE33).unwrap();
+
+    let pmt = PartialMerkleTree::with_paths([(3_u64, path33.value.into(), path33.path)]).unwrap();
+
+    assert_eq!(ms.get_node(expected_root, NODE32).unwrap(), *pmt.get_node(NODE32).unwrap());
+    assert_eq!(ms.get_node(expected_root, NODE10).unwrap(), *pmt.get_node(NODE10).unwrap());
+}
+
+#[test]
+fn update_leaf() {
```
Comment (bobbinth) [inline src/merkle/partial_mt/tests.rs:86]: I would add a couple more tests to this test - i.e., updating leaves at different depths.

File: src/merkle/partial_mt/tests.rs [hunk 25]
```diff
@@ -0,0 +1,128 @@
...
+
+    let path33 = ms.get_path(expected_root, NODE33).unwrap();
+
+    let pmt = PartialMerkleTree::with_paths([(3_u64, path33.value.into(), path33.path)]).unwrap();
+
+    assert_eq!(pmt.root(), expected_root.into());
+}
+
+#[test]
+fn add_and_get_paths() {
```
Comment (bobbinth) [inline src/merkle/partial_mt/tests.rs:48]: I think this is more about testing `with_paths()` constructor rather than `add_path()`.

File: src/merkle/partial_mt/tests.rs [hunk 26]
```diff
@@ -0,0 +1,128 @@
...
+    assert_eq!(pmt.get_leaf_depth(2), 3);
+    assert_eq!(pmt.get_leaf_depth(3), 3);
+    assert_eq!(pmt.get_leaf_depth(4), 1);
+    assert_eq!(pmt.get_leaf_depth(5), 1);
+    assert_eq!(pmt.get_leaf_depth(6), 1);
+    assert_eq!(pmt.get_leaf_depth(7), 1);
+}
+
+// TODO: add test for add_path function and check correctness of leaf determination (requires
+// inner_nodes iter)
```
Comment (bobbinth) [inline src/merkle/partial_mt/tests.rs:128]: I think we should still be able to test `add_path()` separately here but incrementally adding path the a PMT and making sure the root changes as expected.

Also, would be great to add check for `leaves()` in many of these tests, and maybe create a new test for checking `get_paths()`.

**Review by bobbinth — CHANGES_REQUESTED** [review]
Looks good! Thank you! I left a few more comments inline.

File: src/merkle/partial_mt/mod.rs (hunk 21 above)
Comment (Overcastan) [reply src/merkle/partial_mt/mod.rs:108]: With this approach we will have difficulties with calculating value of the node index. Since `move_up_to` function changes not only depth, but also value, dividing it by `2^(max_depth - actual_depth)`, we need to calculate appropriate value for the node index with max depth. So the value of the `NodeIndex` with `depth = 64` will be 
`index * 2^(max_depth - actual_depth)`. So this approach will require calculating the value twice. That's not a big deal, but is it worth it?

Comment (bobbinth) [reply src/merkle/partial_mt/mod.rs:108]: I'm not sure I follow. I was thinking we could do something like:
```Rust
let mut node_index = NodeIndex::new(64, index).expect("");
node_index.move_up_to(self.max_depth());
```
Or is this not sufficient for some reason?

Comment (bobbinth) [reply src/merkle/partial_mt/mod.rs:108]: Ah - I see! I wonder if we should re-think this method somewhat. Maybe the `index` it takes should already be `NodeIndex` which specifies both value and depth. Otherwise, it seems like we may be getting unintuitive results from it.

File: src/merkle/partial_mt/tests.rs [hunk 27]
```diff
@@ -0,0 +1,224 @@
...
+    let root = mt.root();
+
+    let ms = MerkleStore::from(&mt);
+    let path33 = ms.get_path(root, NODE33).unwrap();
+
+    let mut pmt = PartialMerkleTree::with_paths([(3, path33.value.into(), path33.path)]).unwrap();
+
+    let new_value32 = int_to_node(132);
+    mt.update_leaf(2, new_value32).unwrap();
+    let expected_root = mt.root();
```
Comment (bobbinth) [inline src/merkle/partial_mt/tests.rs:102]: Is there a reason to update the Merkle tree rather than Merkle store? I was thinking of Merkle tree as just a way to initialize the Merkle store and then all updates would be done to Merkle store (this should simplify the tests).

File: src/merkle/partial_mt/mod.rs [hunk 28]
```diff
@@ -0,0 +1,332 @@
...
+            return Err(MerkleError::DepthTooSmall(depth));
+        } else if Self::MAX_DEPTH < depth {
+            return Err(MerkleError::DepthTooBig(depth as u64));
+        }
+        Ok(())
+    }
+}
+
+/// Utility to visualize a [PartialMerkleTree] in text.
+pub fn pmt_to_text(tree: &PartialMerkleTree) -> Result<String, fmt::Error> {
```
Comment (bobbinth) [inline src/merkle/partial_mt/mod.rs:306]: If we want to expose this publicly, I'd probably put this as a method on the `PartialMerkleTree` itself. Something like:
```Rust
pub fn print(&self) -> Result<String, fmt::Error>
```

File: src/merkle/partial_mt/mod.rs [hunk 29]
```diff
@@ -0,0 +1,332 @@
...
+    pub fn get_leaf_depth(&self, node_index: NodeIndex) -> u8 {
+        let mut node_index = node_index;
+        for _ in 0..node_index.depth() {
+            if self.leaves.contains(&node_index) {
+                return node_index.depth();
+            }
+            node_index.move_up()
+        }
+        0
+    }
```
Comment (bobbinth) [inline src/merkle/partial_mt/mod.rs:113]: The more I think about it, the less I'm sure about the specifics of this method. Let's remove it for now, and we can re-introduce it later once we get a better sense of what it could be used for.

**Review by bobbinth — CHANGES_REQUESTED** [review]
Thank you! Looks good. I left a few more comments inline. Also, a couple of general comments about tests:

- When a test has several things that it is doing, it would be good to have a brief comment explaining what is being tested.
- We should probably add tests for error conditions. For example, adding two Merkle paths which resolve to different roots should be an error.

**Review by bobbinth — APPROVED** [review]
All looks good! Thank you!

For the next PR, let's add some additional methods (e.g., `inner_nodes()`) and anything else that we'd need to make this a drop-in replacement for `MerklePathSet`.
